File: backend/services/palette.py
"""Palette helpers for preview and pin rendering."""

from __future__ import annotations

from io import BytesIO
from typing import Any

import requests


def normalize_hex_color(value: str | None, default: str) -> str:
    raw = (value or "").strip()
    if not raw:
        return default
    if not raw.startswith("#"):
        raw = f"#{raw}"
    if len(raw) == 4:
        raw = "#" + "".join(ch * 2 for ch in raw[1:])
    if len(raw) != 7:
        return default
    try:
        int(raw[1:], 16)
        return raw.lower()
    except ValueError:
        return default


def _hex_to_rgb(value: str) -> tuple[int, int, int]:
    normalized = normalize_hex_color(value, "#000000")
    return (
        int(normalized[1:3], 16),
        int(normalized[3:5], 16),
        int(normalized[5:7], 16),
    )


def _rgb_to_hex(rgb: tuple[int, int, int]) -> str:
    return "#{:02x}{:02x}{:02x}".format(*[max(0, min(255, int(channel))) for channel in rgb])


def _mix_rgb(
    start: tuple[int, int, int],
    end: tuple[int, int, int],
    ratio: float,
) -> tuple[int, int, int]:
    amount = max(0.0, min(1.0, ratio))
    return tuple(
        int(round(start[index] + (end[index] - start[index]) * amount))
        for index in range(3)
    )


def _luminance(rgb: tuple[int, int, int]) -> float:
    def channel(value: int) -> float:
        scaled = value / 255
        if scaled <= 0.03928:
            return scaled / 12.92
        return ((scaled + 0.055) / 1.055) ** 2.4

    red, green, blue = rgb
    return 0.2126 * channel(red) + 0.7152 * channel(green) + 0.0722 * channel(blue)


def _contrast_ratio(a: tuple[int, int, int], b: tuple[int, int, int]) -> float:
    lighter = max(_luminance(a), _luminance(b))
    darker = min(_luminance(a), _luminance(b))
    return (lighter + 0.05) / (darker + 0.05)


def derive_palette_from_hex(base_color: str) -> dict[str, str]:
    """Create a readable text palette from a dominant image color."""
    base_rgb = _hex_to_rgb(base_color)
    base_luminance = _luminance(base_rgb)
    white = (255, 255, 255)
    black = (17, 24, 39)

    surface_mix = 0.78 if base_luminance < 0.55 else 0.58
    surface_rgb = _mix_rgb(base_rgb, white, surface_mix)

    dark_candidate = _mix_rgb(base_rgb, black, 0.78)
    light_candidate = _mix_rgb(base_rgb, white, 0.9)
    text_rgb = (
        dark_candidate
        if _contrast_ratio(dark_candidate, surface_rgb) >= _contrast_ratio(light_candidate, surface_rgb)
        else light_candidate
    )
    if _contrast_ratio(text_rgb, surface_rgb) < 4.2:
        text_rgb = black if _contrast_ratio(black, surface_rgb) >= _contrast_ratio(white, surface_rgb) else white

    effect_target = white if _luminance(text_rgb) < 0.45 else black
    effect_rgb = _mix_rgb(base_rgb, effect_target, 0.45)
    if _contrast_ratio(effect_rgb, text_rgb) < 1.25:
        effect_rgb = effect_target

    return {
        "background": _rgb_to_hex(surface_rgb),
        "text": _rgb_to_hex(text_rgb),
        "effect": _rgb_to_hex(effect_rgb),
    }


def sample_image_palette(
    image_url: str | None,
    *,
    referer: str | None = None,
    timeout: int = 15,
) -> dict[str, str] | None:
    """Download an image and derive a lightweight palette from it."""
    if not image_url:
        return None

    try:
        from PIL import Image
    except Exception:
        return None

    try:
        response = requests.get(
            image_url,
            timeout=timeout,
            headers={
                "User-Agent": "Mozilla/5.0 (compatible; PinterestCSVTool/1.0)",
                "Accept": "image/avif,image/webp,image/apng,image/*,*/*;q=0.8",
                "Referer": referer or "",
            },
        )
        response.raise_for_status()
        with Image.open(BytesIO(response.content)) as image:
            sampled = image.convert("RGB")
            sampled.thumbnail((24, 24))
            pixels = list(sampled.getdata())
            if not pixels:
                return None
            average = tuple(
                int(round(sum(pixel[index] for pixel in pixels) / len(pixels)))
                for index in range(3)
            )
            return derive_palette_from_hex(_rgb_to_hex(average))
    except Exception:
        return None


def resolve_palette_settings(
    base_settings: dict[str, Any],
    *,
    image_url: str | None = None,
    referer: str | None = None,
) -> dict[str, Any]:
    """Resolve palette mode into concrete render colors."""
    resolved = dict(base_settings)
    palette_mode = (resolved.get("palette_mode") or "").strip().lower()
    current_background = normalize_hex_color(resolved.get("text_zone_bg_color"), "#ffffff")
    current_text = normalize_hex_color(resolved.get("text_color"), "#000000")
    current_effect = normalize_hex_color(resolved.get("text_effect_color"), "#000000")

    if palette_mode == "brand":
        resolved["text_zone_bg_color"] = normalize_hex_color(
            resolved.get("brand_palette_background_color"),
            current_background,
        )
        resolved["text_color"] = normalize_hex_color(
            resolved.get("brand_palette_text_color"),
            current_text,
        )
        resolved["text_effect_color"] = normalize_hex_color(
            resolved.get("brand_palette_effect_color"),
            current_effect,
        )
        return resolved

    if palette_mode == "manual":
        resolved["text_zone_bg_color"] = normalize_hex_color(
            resolved.get("manual_palette_background_color"),
            current_background,
        )
        resolved["text_color"] = normalize_hex_color(
            resolved.get("manual_palette_text_color"),
            current_text,
        )
        resolved["text_effect_color"] = normalize_hex_color(
            resolved.get("manual_palette_effect_color"),
            current_effect,
        )
        return resolved

    if palette_mode == "auto":
        palette = sample_image_palette(image_url, referer=referer)
        if palette:
            resolved["text_zone_bg_color"] = palette["background"]
            resolved["text_color"] = palette["text"]
            resolved["text_effect_color"] = palette["effect"]
            return resolved

    resolved["text_zone_bg_color"] = current_background
    resolved["text_color"] = current_text
    resolved["text_effect_color"] = current_effect
    return resolved

File: frontend/src/utils/palette.ts
export type PaletteMode = 'auto' | 'brand' | 'manual';

export type EditablePalette = {
  background: string;
  text: string;
  effect: string;
};

export function normalizeHexColor(value: string | null | undefined, fallback: string): string {
  const raw = (value || '').trim();
  if (!raw) return fallback;
  const prefixed = raw.startsWith('#') ? raw : `#${raw}`;
  if (/^#[0-9a-f]{3}$/i.test(prefixed)) {
    return `#${prefixed[1]}${prefixed[1]}${prefixed[2]}${prefixed[2]}${prefixed[3]}${prefixed[3]}`.toLowerCase();
  }
  if (!/^#[0-9a-f]{6}$/i.test(prefixed)) return fallback;
  return prefixed.toLowerCase();
}

function hexToRgb(value: string): [number, number, number] {
  const normalized = normalizeHexColor(value, '#000000');
  return [
    Number.parseInt(normalized.slice(1, 3), 16),
    Number.parseInt(normalized.slice(3, 5), 16),
    Number.parseInt(normalized.slice(5, 7), 16),
  ];
}

function rgbToHex([r, g, b]: [number, number, number]): string {
  return `#${[r, g, b].map((channel) => Math.max(0, Math.min(255, Math.round(channel))).toString(16).padStart(2, '0')).join('')}`;
}

function mixRgb(
  start: [number, number, number],
  end: [number, number, number],
  ratio: number,
): [number, number, number] {
  const amount = Math.max(0, Math.min(1, ratio));
  return [
    start[0] + (end[0] - start[0]) * amount,
    start[1] + (end[1] - start[1]) * amount,
    start[2] + (end[2] - start[2]) * amount,
  ] as [number, number, number];
}

function luminance([r, g, b]: [number, number, number]): number {
  const channel = (value: number) => {
    const scaled = value / 255;
    return scaled <= 0.03928 ? scaled / 12.92 : ((scaled + 0.055) / 1.055) ** 2.4;
  };
  return 0.2126 * channel(r) + 0.7152 * channel(g) + 0.0722 * channel(b);
}

function contrastRatio(a: [number, number, number], b: [number, number, number]): number {
  const lighter = Math.max(luminance(a), luminance(b));
  const darker = Math.min(luminance(a), luminance(b));
  return (lighter + 0.05) / (darker + 0.05);
}

export function derivePaletteFromHex(baseColor: string): EditablePalette {
  const baseRgb = hexToRgb(baseColor);
  const baseLuminance = luminance(baseRgb);
  const white: [number, number, number] = [255, 255, 255];
  const black: [number, number, number] = [17, 24, 39];

  const surfaceMix = baseLuminance < 0.55 ? 0.78 : 0.58;
  const surfaceRgb = mixRgb(baseRgb, white, surfaceMix);

  const darkCandidate = mixRgb(baseRgb, black, 0.78);
  const lightCandidate = mixRgb(baseRgb, white, 0.9);
  let textRgb = contrastRatio(darkCandidate, surfaceRgb) >= contrastRatio(lightCandidate, surfaceRgb)
    ? darkCandidate
    : lightCandidate;
  if (contrastRatio(textRgb, surfaceRgb) < 4.2) {
    textRgb = contrastRatio(black, surfaceRgb) >= contrastRatio(white, surfaceRgb) ? black : white;
  }

  const effectTarget = luminance(textRgb) < 0.45 ? white : black;
  let effectRgb = mixRgb(baseRgb, effectTarget, 0.45);
  if (contrastRatio(effectRgb, textRgb) < 1.25) {
    effectRgb = effectTarget;
  }

  return {
    background: rgbToHex(surfaceRgb),
    text: rgbToHex(textRgb),
    effect: rgbToHex(effectRgb),
  };
}

export async function sampleImagePalette(imageUrl: string): Promise<EditablePalette | null> {
  if (!imageUrl) return null;
  return new Promise((resolve) => {
    const image = new Image();
    image.crossOrigin = 'anonymous';
    image.onload = () => {
      try {
        const canvas = document.createElement('canvas');
        canvas.width = 24;
        canvas.height = 24;
        const ctx = canvas.getContext('2d');
        if (!ctx) {
          resolve(null);
          return;
        }
        ctx.drawImage(image, 0, 0, canvas.width, canvas.height);
        const pixels = ctx.getImageData(0, 0, canvas.width, canvas.height).data;
        let red = 0;
        let green = 0;
        let blue = 0;
        let count = 0;
        for (let index = 0; index < pixels.length; index += 4) {
          const alpha = pixels[index + 3];
          if (alpha < 8) continue;
          red += pixels[index];
          green += pixels[index + 1];
          blue += pixels[index + 2];
          count += 1;
        }
        if (!count) {
          resolve(null);
          return;
        }
        const average = rgbToHex([red / count, green / count, blue / count] as [number, number, number]);
        resolve(derivePaletteFromHex(average));
      } catch {
        resolve(null);
      }
    };
    image.onerror = () => resolve(null);
    image.src = imageUrl;
  });
}
